sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java

```diff
@@ -235,12 +284,151 @@ public ReadFiles withAvroDataModel(GenericData model) {
+          OutputReceiver<GenericRecord> outputReceiver)
+          throws Exception {
+        if (!file.getMetadata().isReadSeekEfficient()) {
+          ResourceId filename = file.getMetadata().resourceId();
+          throw new RuntimeException(String.format("File has to be seekable: %s", filename));
+        }
+
+        SeekableByteChannel seekableByteChannel = file.openSeekable();
+        ReadSupport<GenericRecord> readSupport;
+        InputFile inputFile = new BeamParquetInputFile(seekableByteChannel);
+        Configuration conf = setConf();
+        GenericData model = null;
+        if (modelClass != null) {
+          model = (GenericData) modelClass.getMethod("get").invoke(null);
+        }
+        readSupport = new AvroReadSupport<GenericRecord>(model);
+        ParquetReadOptions options = HadoopReadOptions.builder(conf).build();
+        ParquetFileReader reader = ParquetFileReader.open(inputFile, options);
+        Filter filter = checkNotNull(options.getRecordFilter(), "filter");
+        conf = ((HadoopReadOptions) options).getConf();
+        for (String property : options.getPropertyNames()) {
+          conf.set(property, options.getProperty(property));
+        }
+        FileMetaData parquetFileMetadata = reader.getFooter().getFileMetaData();
+        MessageType fileSchema = parquetFileMetadata.getSchema();
+        Map<String, String> fileMetadata = parquetFileMetadata.getKeyValueMetaData();
+
+        ReadSupport.ReadContext readContext =
+            readSupport.init(new InitContext(conf, toSetMultiMap(fileMetadata), fileSchema));
+        ColumnIOFactory columnIOFactory = new ColumnIOFactory(parquetFileMetadata.getCreatedBy());
+        MessageType requestedSchema = readContext.getRequestedSchema();
+        RecordMaterializer<GenericRecord> recordConverter =
+            readSupport.prepareForRead(conf, fileMetadata, fileSchema, readContext);
+        boolean strictTypeChecking = options.isEnabled(STRICT_TYPE_CHECKING, true);
+        boolean filterRecords = options.useRecordFilter();
+        reader.setRequestedSchema(requestedSchema);
+        MessageColumnIO columnIO =
+            columnIOFactory.getColumnIO(requestedSchema, fileSchema, strictTypeChecking);
+        for (int i = 0; i < reader.getRowGroups().size(); i++) {
+          if (i < tracker.currentRestriction().getFrom()) {
+            reader.skipNextRowGroup();
+            continue;
+          }
+          if (tracker.tryClaim((long) i)) {
+            PageReadStore pages = reader.readNextRowGroup();
+            i += 1;
+            RecordReader<GenericRecord> recordReader =
+                columnIO.getRecordReader(
+                    pages, recordConverter, filterRecords ? filter : FilterCompat.NOOP);
+            GenericRecord read;
+            long current = 0;
+            long totalRows = pages.getRowCount();
+            while (current < totalRows) {
+              read = recordReader.read();
+              outputReceiver.output(read);
```

Comment:
You could do ` outputReceiver.output(recordReader.read())` to get rid of local var.